core/src/main/java/com/linecorp/armeria/common/stream/InputStreamStreamMessage.java

```diff
@@ -0,0 +1,384 @@
+        if (completionFuture.isCompletedExceptionally()) {
+            completionFuture.exceptionally(cause -> {
+                inputStreamSubscription.close(cause);
+                return null;
+            });
+        }
+    }
+
+    @Override
+    public void abort() {
+        abort(AbortedStreamException.get());
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+
+        // `completionFuture` should be set before `inputStreamSubscription` is read
+        // to guarantee the visibility of the abortion `cause` after
+        // inputStreamSubscription is set in `subscriber0()`.
+        completionFuture.completeExceptionally(cause);
+
+        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
+        if (inputStreamSubscription != null) {
+            inputStreamSubscription.close(cause);
+        }
+    }
+
+    private final class InputStreamSubscription implements Subscription {
+
+        private Subscriber<? super HttpData> downstream;
+        private final EventExecutor executor;
+        private final ExecutorService blockingTaskExecutor;
+        private final int bufferSize;
+
+        private final long offset;
+        private final long end;
+        private long position;
+        private volatile boolean written;
+
+        private final boolean notifyCancellation;
+
+        private boolean closed;
+        private volatile long requested;
+        private boolean reading;
+
+        private InputStreamSubscription(Subscriber<? super HttpData> downstream, EventExecutor executor,
+                                        ExecutorService blockingTaskExecutor, int bufferSize,
+                                        long offset, long length, boolean notifyCancellation) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(executor, "executor");
+            requireNonNull(blockingTaskExecutor, "blockingTaskExecutor");
+            this.downstream = downstream;
+            this.executor = executor;
+            this.blockingTaskExecutor = blockingTaskExecutor;
+            this.bufferSize = bufferSize;
+            this.offset = offset;
+            end = LongMath.saturatedAdd(offset, length);
+            this.notifyCancellation = notifyCancellation;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0L) {
+                close(new IllegalArgumentException(
+                        "Rule §3.9 violated: non-positive subscription requests are forbidden."));
+                return;
+            }
+
+            if (executor.inEventLoop()) {
+                request0(n);
+            } else {
+                executor.execute(() -> request0(n));
+            }
+        }
+
+        private void request0(long n) {
+            if (closed) {
+                return;
+            }
+            if (offset >= end) {
+                close(null);
```

Comment:
Safely call `close0()`
```suggestion
                close0(null);
```